File: mcspring-api/mcspring-base/src/main/java/in/kyle/mcspring/command/registration/BukkitCommandRegistration.java
package in.kyle.mcspring.command.registration;

import org.springframework.stereotype.Component;

import java.lang.reflect.Method;
import java.util.Arrays;

import in.kyle.mcspring.RequiresSpigot;
import in.kyle.mcspring.command.Command;
import lombok.RequiredArgsConstructor;
import lombok.var;

@Component
@RequiredArgsConstructor
@RequiresSpigot
class BukkitCommandRegistration implements CommandRegistration {
    
    private final CommandController controller;
    private final CommandFactory commandFactory;
    
    @Override
    public void register(Command command, Method method, Object object) {
        String name = command.value();
        var bukkitCommand = commandFactory.makeCommand(method, object, name);
        bukkitCommand.setAliases(Arrays.asList(command.aliases()));
        bukkitCommand.setDescription(command.description());
        bukkitCommand.setUsage(command.usage());
        bukkitCommand.setPermission(command.permission());
        bukkitCommand.setPermissionMessage(command.permissionMessage());
        controller.registerCommand(bukkitCommand);
    }
    
    interface CommandFactory {
        org.bukkit.command.Command makeCommand(Method method, Object object, String name);
    }
}

File: mcspring-api/mcspring-base/src/main/kotlin/in/kyle/mcspring/command/registration/BukkitCommandRegistration.kt
package `in`.kyle.mcspring.command.registration

import `in`.kyle.mcspring.command.Command
import org.bukkit.plugin.Plugin
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean
import org.springframework.stereotype.Component
import java.lang.reflect.Method

@Component
@ConditionalOnBean(Plugin::class)
internal class BukkitCommandRegistration(
        private val controller: CommandController,
        private val commandExecutorFactory: CommandExecutorFactory
) : CommandRegistration {

    override fun register(command: Command, method: Method, obj: Any) {
        val bukkitCommand = commandExecutorFactory.makeCommand(method, obj, command.value)
        with(bukkitCommand) {
            aliases = command.aliases.toList()
            description = command.description
            usage = command.usage
            permission = command.permission
            permissionMessage = command.permissionMessage
        }
        controller.registerCommand(bukkitCommand)
    }
}
